core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -912,12 +969,9 @@ class LogManager(logDirs: Seq[File],
     if (removedLog != null) {
       //We need to wait until there is no more cleaning task on the log to be deleted before actually deleting it.
       if (cleaner != null && !isFuture) {
-        cleaner.abortCleaning(topicPartition)
-        cleaner.updateCheckpoints(removedLog.parentDirFile)
+        cleaner.abortCleaning(topicPartition, partitionDeleted = true)
       }
       removedLog.renameDir(Log.logDeleteDirName(topicPartition))
-      checkpointRecoveryOffsetsAndCleanSnapshot(removedLog.parentDirFile, ArrayBuffer.empty)
```

Comment:
With this change, it is up to the caller to ensure offsets are checkpointed correctly after a deletion. It would be better from an encapsulation perspective to keep `LogManager` in charge of that. One idea would be to offer a batched `asyncDelete`. The involvement of `Partition` may make this a bit complex though.

Reply:
I totally agree with you. I have tried different ways trying to keep a better encapsulation but I have found a really satisfying way to get there. As you pointed out, the involvement of `Partition` makes this complex.

Here is my best idea so far:
* We remove the calls to `asyncDelete` in the `Partition.delete` method. It seems safe to "delete the partition" while keeping the files on disk while holding the `replicaStateChangeLock` in the `ReplicaManager`.
* We use a `asyncDelete` that takes a batch of logs, deletes them and checkpoint.
* We rename `Partition.delete` to something like `Partition.close` as the log is not really deleted any more.

What do you think?

Reply:
I have implemented it to see. You can check out the last commit.